### PR TITLE
fixed compare and added tests

### PR DESCRIPTION
**Pull request by timothycobrien**
Fixing #78 , also modified .gitignore for my IDE

### CODE REVIEW

**Comment by luke-bhan** [comment]
ok, depending on what @ayazhafiz thinks, maybe we should just edit the constructor, I didn't think this would be that kind of problem. I'm sorry. It is just a one liner if we do that so something like what we do with the dec vector:
```
if !int.is_empty() && int.iter().all(|&i| i == 0) { 
    int.clear();
}
```
Then just I would just check compare.rs if all vectors are empty and return equal.
(Also, I think the less than error comes from a switching of orders in the match statement on the last line of compare.rs) 

I think that would shorten this commit to 6 or 7 lines. And I don't think we will run into integration tests since all the operators right now strip the leading zeros which would work for a vector of all zeros. What do you think? @timothycobrien

**Comment by ayazhafiz** [comment]
@luke-bhan do all operations drop trailing zeros on the decimal as well?

Explicitly dropping zeros everywhere is fine, but if we do that let's document that we're doing so on the bignum struct.

File: libslide/src/bignum/compare.rs [hunk 1]
```diff
@@ -3,53 +3,104 @@ use std::cmp::Ordering;
 
 impl Ord for Bignum {
     fn cmp(&self, other: &Bignum) -> Ordering {
+        let (lhs, rhs) = (strip_zeros(self), strip_zeros(other));
+        // lhs and rhs have no leading or trailing (in decimal portion) zeros
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:7]: There's no need to allocate a new bignum; just keep a small int pointing to where leading/trailing zeros (int/decimal) stand.

File: libslide/src/bignum/compare.rs [hunk 2]
```diff
@@ -3,53 +3,104 @@ use std::cmp::Ordering;
...
         let greater: Ordering;
         let less: Ordering;
-        if self.is_neg ^ other.is_neg {
-            return other.is_neg.cmp(&self.is_neg);
-        } else if other.is_neg && self.is_neg {
+        if lhs.is_neg ^ rhs.is_neg { // need to check for negative 0
+            // use the fact that a pure zero will have an empty int and empty dec vector
+            if lhs.dec.is_empty() && rhs.dec.is_empty() && lhs.int.is_empty() && rhs.int.is_empty() {
+                return Ordering::Equal;
+            }
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:14]: then this is equivalent to checking `lhs_dec_term == 0 && rhs_dec_term == lhs_dec_term && ...`

**Comment by luke-bhan** [comment]
theyre supposed to. I’ll add tests in a separate pr

**Comment by ayazhafiz** [comment]
Okay, modifying the ctor sounds good to me

File: libslide/src/bignum/compare.rs (hunk 1 above)
Comment (timothycobrien) [reply libslide/src/bignum/compare.rs:7]: Do you mean not modifying either of the Vecs at all and just altering ranges for int-by-int checks?

**Comment by ayazhafiz** [comment]
yes. but luke bhan's idea is probably the way to go

**Comment by timothycobrien** [comment]
> Okay, modifying the ctor sounds good to me

I think it makes sense to @ayazhafiz, unless we're specifically making a distinction between *1* and *1.0*, but I don't think that's the case here? I think it makes sense to just enforce a non-empty int Vec and a non-empty dec Vec, so even *0* would be represented as *0.0*. 

This provides a consistent structure which should make future functions much easier to implement.

**Comment by ayazhafiz** [comment]
I'm not sure I see the benefit of `0.0` vs `.` where the int/dec vectors are empty. Can you elaborate with an example where this makes a difference?

**Comment by timothycobrien** [comment]
I didn't necessarily have a specific issue in mind, but doing something like `0+1` where `0` is actually `.` versus `0.0` *seems* more complicated to me, as that edge case has to be coded in versus being handled like any other case.

**Comment by timothycobrien** [comment]
I just thought of this, but it might make sense to modify all of the comparisons entirely so a comparison of `x > y` amounts to `x - y > 0`,  `x < y` amounts to `y - x > 0`, so we can just test if a number is negative? This is actually where having `0` be represented by empty dec and int vectors might come in handy, but not sure if that alone is worth it imo.

**Comment by ayazhafiz** [comment]
> I just thought of this, but it might make sense to modify all of the comparisons entirely so a comparison of `x > y` amounts to `x - y > 0`, `x < y` amounts to `y - x > 0`, so we can just test if a number is negative?

This would need benchmarking but I my hypothesis is it won't be faster.

**Comment by timothycobrien** [comment]
Having a consistent structure with two non empty vectors seems much better to me for future extensibility than the possibility of one being empty, but I'm having a tough time coming up with a concrete reason why. Thoughts @luke-bhan ?

**Comment by timothycobrien** [comment]
> > I just thought of this, but it might make sense to modify all of the comparisons entirely so a comparison of `x > y` amounts to `x - y > 0`, `x < y` amounts to `y - x > 0`, so we can just test if a number is negative?
> 
> This would need benchmarking but I my hypothesis is it won't be faster.

Was mostly thinking in terms of re-usability here, would also be easier to modify later on (somehow providing a different binary relation than the typical `<, >, =` or modifying their meanings)

**Comment by luke-bhan** [comment]
I see where you're coming from and I don't disagree because it does seem more intuitive. I just don't know how we would handle operations like 0.00 and 0.0000 in the add method. Would the result just be 0.0 so we would have to truncate everything except the last zero? I think its just easier to write a method to trim all vectors (truncate_vectros) and just use that.

**Comment by ayazhafiz** [comment]
I don’t think keeping one zero on both sides is consistent with dropping leading and trailing zeros.

**Comment by timothycobrien** [comment]
Fair point Ayaz, I think it's a matter mostly of convention at this point. Obviously we write `0` on paper, because having `.`  on a paper doesn't make any sense, but no reason why we couldn't just have `.` in the code besides I guess human readability?

**Comment by ayazhafiz** [comment]
yeah, i agree that for the human `0.0` is better. For that the to_string can be modified.

**Comment by timothycobrien** [comment]
This should be good to go now @ayazhafiz, didn't fix the to_string or anything yet, but all the compare tests pass now

File: libslide/src/bignum/compare.rs [hunk 3]
```diff
@@ -6,6 +6,13 @@ impl Ord for Bignum {
         let greater: Ordering;
         let less: Ordering;
         if self.is_neg ^ other.is_neg {
+            /* need to check for negative 0
+            use the fact that a pure zero will have an empty int and empty dec vector
+            if all the lengths sum to zero, each length must itself be 0
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:11]: what if one length is negative ;)

Comment (timothycobrien) [reply libslide/src/bignum/compare.rs:11]: Troll

**Comment by timothycobrien** [comment]
Any idea on why CI isn't passing @ayazhafiz ?

**Comment by ayazhafiz** [comment]
Shoot, it’s because the squash script expects the branch to be on origin... I’ll fix it after work today, or you can take a go at it

**Comment by timothycobrien** [comment]
> Shoot, it’s because the squash script expects the branch to be on origin... I’ll fix it after work today, or you can take a go at it

I'll take a look

**Comment by timothycobrien** [comment]
Ok nevermind @ayazhafiz, I thought I knew what I was doing vis-a-vis Travis, but I have no idea haha.

Is it not just modifying `.travis.yml` to have this line instead ` env: BRANCH="$TRAVIS_REPO_SLUG/$TRAVIS_PULL_REQUEST_BRANCH"`? Honestly, if you could explain the actual fix to me that would be awesome.

**Comment by ayazhafiz** [comment]
@timothycobrien you're on the right track.

1. You need to get the repo slug and concat it with GitHub url to get a path to a branch. For example
in this case "https://github.com/timothycobrien/slide". Make this a new environment variable or argument for `scripts/squashed`, maybe TARGET_BRANCH_REPO if its an env variable

2. This should be used to add a remote in the same way the `squash_check` remote is set: https://github.com/ayazhafiz/scripts/blob/4fc2ca6217282ef5be9bacb7bcd71e46bfad2781/git/squashed#L29-L30

3. Then, use the branch of the remote in place of the raw BRANCH env variable (https://github.com/ayazhafiz/scripts/blob/4fc2ca6217282ef5be9bacb7bcd71e46bfad2781/git/squashed#L33). Specifically, this should be changed to

```bash
BRANCH="${TARGET_BRANCH_REMOTE}/${BRANCH:-$(git branch --show-current)}"
```

Let me know if this makes sense or if I can help more.

**Comment by luke-bhan** [comment]
@timothycobrien is it ok if I just fix and merge this?

**Comment by timothycobrien** [comment]
I'm looking at it now @luke-bhan, but honestly really new to Travis, so I'll ping you in 30 if I can't make any progress.

**Comment by timothycobrien** [comment]
@ayazhafiz what do the `//` mean on line 33? Also, not quite sure if everything is correct, @luke-bhan please help lol

**Comment by luke-bhan** [comment]
I think you need to change line 37 to
```
BRANCH="${TARGET_BRANCH_REMOTE}/${BRANCH:-$(git branch --show-current)}"
```

**Comment by timothycobrien** [comment]
Still doesn't work @luke-bhan :(

**Comment by luke-bhan** [comment]
oh wait also you need to change the repo env variable to your repo so github/timothycobrien/slide because it is still set to ayazhafiz/slide. There may be other problems but try that.

**Comment by luke-bhan** [comment]
hmm somewhere we have an invalid variable. I would say just echo repo, target_branch_Remote, branch, etc. and it ideally will show where we have the invalid arg

**Comment by luke-bhan** [comment]
instead of dealing with this, I externalized bignum and then updated it. It is now available at https://github.com/luke-bhan/numlib if you are still interested in fixing the fft/implementing new algos. The travis and scripts are not updated yet though. @timothycobrien

**Comment by timothycobrien** [comment]
I'd like to try the other multiplication algo and I have a viable (though hardly complete or optimized) division algo already written but haven't been able to test it without multiplication working correctly.

**Pull request closed without merging**